documents/CONFIG.md
<!-- TOC -->

- [Basic](#basic)
    - [TRAVELLING_PORT](#travelling_port)
    - [TRAVELLING_IP](#travelling_ip)
    - [TRAVELLING_KEY](#travelling_key)
    - [TRAVELLING_CERT](#travelling_cert)
    - [TRAVELLING_HTTPS](#travelling_https)
    - [TRAVELLING_SERVICE_NAME](#travelling_service_name)
- [Misc](#misc)
    - [TRAVELLING_MISC_CLOUDFLAREIP](#travelling_misc_cloudflareip)
- [Cors](#cors)
    - [TRAVELLING_CORS_ENABLE](#travelling_cors_enable)
    - [TRAVELLING_CORS_HEADER_ORIGIN](#travelling_cors_header_origin)
    - [TRAVELLING_CORS_HEADER_METHODS](#travelling_cors_header_methods)
    - [TRAVELLING_CORS_HEADER_CREDENTIALS](#travelling_cors_header_credentials)
    - [TRAVELLING_CORS_HEADER_MAX_AGE](#travelling_cors_header_max_age)
- [Logging](#logging)
    - [TRAVELLING_LOG_ENABLE](#travelling_log_enable)
    - [TRAVELLING_LOG_LOGGER](#travelling_log_logger)
    - [TRAVELLING_LOG_COLORS](#travelling_log_colors)
    - [TRAVELLING_LOG_LEVEL](#travelling_log_level)
    - [TRAVELLING_LOG_REQUESTS](#travelling_log_requests)
    - [TRAVELLING_LOG_UNAUTHORIZED_ACCESS](#travelling_log_unauthorized_access)
- [Fastify Logger](#fastify-logger)
    - [TRAVELLING_LOG_FASTIFY_LOGGER](#travelling_log_fastify_logger)
    - [TRAVELLING_LOG_FASTIFY_LOGGER_REQUEST](#travelling_log_fastify_logger_request)
    - [TRAVELLING_LOG_FASTIFY_LOGGER_REQ_ID_HEADER](#travelling_log_fastify_logger_req_id_header)
    - [TRAVELLING_LOG_FASTIFY_LOGGER_REQ_ID_LOG_LABEL](#travelling_log_fastify_logger_req_id_log_label)
- [Portal](#portal)
    - [TRAVELLING_PORTAL_ENABLE](#travelling_portal_enable)
    - [TRAVELLING_PORTAL_PATH](#travelling_portal_path)
    - [TRAVELLING_PORTAL_HOST](#travelling_portal_host)
    - [TRAVELLING_PORTAL_FILE_PATH](#travelling_portal_file_path)
    - [TRAVELLING_PORTAL_LOGO](#travelling_portal_logo)
    - [TRAVELLING_PORTAL_STYLES](#travelling_portal_styles)
    - [TRAVELLING_PORTAL_ICON](#travelling_portal_icon)
- [Proxy](#proxy)
    - [TRAVELLING_PROXY_TIMEOUT](#travelling_proxy_timeout)
    - [TRAVELLING_PROXY_SEND_TRAVELLING_HEADERS](#travelling_proxy_send_travelling_headers)
- [Redis](#redis)
    - [TRAVELLING_REDIS_ENABLE](#travelling_redis_enable)
    - [TRAVELLING_REDIS_URL](#travelling_redis_url)
    - [TRAVELLING_REDIS_EVENTS_URL](#travelling_redis_events_url)
- [Cookie](#cookie)
    - [TRAVELLING_COOKIE_SESSION_SECRET](#travelling_cookie_session_secret)
    - [TRAVELLING_COOKIE_SESSION_EXPIRATION](#travelling_cookie_session_expiration)
    - [TRAVELLING_COOKIE_TOKEN_SECRET](#travelling_cookie_token_secret)
    - [TRAVELLING_COOKIE_TOKEN_SALT](#travelling_cookie_token_salt)
    - [TRAVELLING_COOKIE_TOKEN_EXPIRATION](#travelling_cookie_token_expiration)
    - [TRAVELLING_COOKIE_TOKEN_CHECKABLE](#travelling_cookie_token_checkable)
    - [TRAVELLING_COOKIE_DOMAIN](#travelling_cookie_domain)
    - [TRAVELLING_COOKIE_SECURITY_IP_HIJACK_PROTECTION](#travelling_cookie_security_ip_hijack_protection)
- [USER](#user)
    - [TRAVELLING_USER_ISOLATE_BY_DOMAIN](#travelling_user_isolate_by_domain)
    - [TRAVELLING_USER_USERNAME_MINCHAR](#travelling_user_username_minchar)
    - [TRAVELLING_USER_USERNAME_ENABLE](#travelling_user_username_enable)
    - [TRAVELLING_USER_LOCKED_MESSAGE](#travelling_user_locked_message)
- [Authentication](#authentication)
    - [TRAVELLING_PASSWORD_CONSECUTIVE](#travelling_password_consecutive)
    - [TRAVELLING_PASSWORD_MINCHAR](#travelling_password_minchar)
    - [TRAVELLING_PASSWORD_MAXCHAR](#travelling_password_maxchar)
    - [TRAVELLING_PASSWORD_SPECIAL](#travelling_password_special)
    - [TRAVELLING_PASSWORD_NUMBER](#travelling_password_number)
    - [TRAVELLING_PASSWORD_LOWERCASE](#travelling_password_lowercase)
    - [TRAVELLING_PASSWORD_UPPERCASE](#travelling_password_uppercase)
    - [TRAVELLING_LOGIN_MAX_LOGIN_ATTEMPTS](#travelling_login_max_login_attempts)
- [OAUTH2](#oauth2)
    - [TRAVELLING_TOKEN_ACCESS_EXPIRATION](#travelling_token_access_expiration)
    - [TRAVELLING_TOKEN_CODE_EXPIRATION](#travelling_token_code_expiration)
    - [TRAVELLING_TOKEN_CODE_AUTHORIZE_FLOW](#travelling_token_code_authorize_flow)
- [Postgres](#postgres)
    - [TRAVELLING_DATABASE_URL](#travelling_database_url)
    - [TRAVELLING_DATABASE_USER](#travelling_database_user)
    - [TRAVELLING_DATABASE_PASSWORD](#travelling_database_password)
    - [TRAVELLING_DATABASE_PORT](#travelling_database_port)
    - [TRAVELLING_DATABASE_NAME](#travelling_database_name)
    - [TRAVELLING_DATABASE_HOST](#travelling_database_host)
    - [TRAVELLING_PG_CRYPTO_IMPLEMENTATION](#travelling_pg_crypto_implementation)
    - [TRAVELLING_PG_CRYPTO_IMPLEMENTATION_SECRET](#travelling_pg_crypto_implementation_secret)
    - [TRAVELLING_PG_CRYPTO_IMPLEMENTATION_SALT](#travelling_pg_crypto_implementation_salt)
    - [TRAVELLING_PG_CRYPTO_ENCRYPT_USER_DATA](#travelling_pg_crypto_encrypt_user_data)
- [Email](#email)
    - [TRAVELLING_EMAIL_VALIDATION_EXTERNAL_ENABLE](#travelling_email_validation_external_enable)
    - [TRAVELLING_EMAIL_VALIDATION_EXTERNAL_ENDPOINT](#travelling_email_validation_external_endpoint)
    - [TRAVELLING_EMAIL_VALIDATION_EXTERNAL_EMAIL_IN_ENDPOINT](#travelling_email_validation_external_email_in_endpoint)
    - [TRAVELLING_EMAIL_VALIDATION_EXTERNAL_EMAIL_IN_BODY](#travelling_email_validation_external_email_in_body)
    - [TRAVELLING_EMAIL_VALIDATION_EXTERNAL_METHOD](#travelling_email_validation_external_method)
    - [TRAVELLING_EMAIL_FROM](#travelling_email_from)
    - [TRAVELLING_EMAIL_RECOVERY_EXPIRATION](#travelling_email_recovery_expiration)
    - [TRAVELLING_EMAIL_ACTIVATION_EXPIRATION](#travelling_email_activation_expiration)
    - [TRAVELLING_EMAIL_TEST_ENABLE](#travelling_email_test_enable)
    - [TRAVELLING_EMAIL_SMTP_ENABLE](#travelling_email_smtp_enable)
    - [TRAVELLING_EMAIL_SMTP_HOST](#travelling_email_smtp_host)
    - [TRAVELLING_EMAIL_SMTP_PORT](#travelling_email_smtp_port)
    - [TRAVELLING_EMAIL_SMTP_SECURE](#travelling_email_smtp_secure)
    - [TRAVELLING_EMAIL_SMTP_AUTH_USER](#travelling_email_smtp_auth_user)
    - [TRAVELLING_EMAIL_SMTP_SECURE](#travelling_email_smtp_secure-1)
    - [TRAVELLING_EMAIL_SMTP_TLS_REJECT_UNAUTHORIZED](#travelling_email_smtp_tls_reject_unauthorized)
    - [TRAVELLING_EMAIL_AWS_ENABLE](#travelling_email_aws_enable)
    - [TRAVELLING_EMAIL_AWS_CONFIG](#travelling_email_aws_config)
  - [Templates](#templates)
    - [TRAVELLING_EMAIL_RESET_PASSWORD_TEMPLATE_BODY](#travelling_email_reset_password_template_body)
    - [TRAVELLING_EMAIL_RESET_PASSWORD_TEMPLATE_SUBJECT](#travelling_email_reset_password_template_subject)
    - [TRAVELLING_EMAIL_ACTIVATION_TEMPLATE_BODY](#travelling_email_activation_template_body)
    - [TRAVELLING_EMAIL_ACTIVATION_TEMPLATE_SUBJECT](#travelling_email_activation_template_subject)
- [Registration](#registration)
    - [TRAVELLING_REGISTRATION_REQUIRE_EMAIL_ACTIVATION](#travelling_registration_require_email_activation)
    - [TRAVELLING_REGISTRATION_REQUIRE_MANUAL_ACTIVATION](#travelling_registration_require_manual_activation)

<!-- /TOC -->

### Basic

##### TRAVELLING_PORT

_Travelling's serving port._ </br>

> **Default**: `443`

##### TRAVELLING_IP

_Travelling's serving IP._ </br>

> **Default**: `0.0.0.0`

##### TRAVELLING_KEY

_The path to the SSL key that is used for [https](#TRAVELLING_HTTPS)_ </br>

> **Default**: `travelling/localhost.key`

##### TRAVELLING_CERT

_The path to the SSL cert that is used for [https](#TRAVELLING_HTTPS)_ </br>


CHANGELOG.md


#### [v2.19.0](https://github.com/Dragohm/travelling/compare/v2.18.1...v2.19.0)

- Feature #66 - Add checkable cookie [`#23`](https://github.com/Dragohm/travelling/pull/23)
- Add ls token tests [`c944378`](https://github.com/Dragohm/travelling/commit/c944378dd4c3bd1c33beaac29e8bdb0a1663d53e)
- Update package.json [`349372a`](https://github.com/Dragohm/travelling/commit/349372aedf5b9feb94072ac4f0e48c0efbe9d3b1)
- update config.md [`69ba80b`](https://github.com/Dragohm/travelling/commit/69ba80bd5f4006ed3cb2ee0f087822718b622916)
- update config.md [`d65c70c`](https://github.com/Dragohm/travelling/commit/d65c70cd5637975c7c0e0ba40e5d4e5c0282df01)
- Add trav:ls cookie [`f0d9770`](https://github.com/Dragohm/travelling/commit/f0d97702af28e62f4157583e5b7208bf60208dc9)

#### [v2.18.1](https://github.com/Dragohm/travelling/compare/v2.18.0...v2.18.1)

> 8 November 2022

- Update auth.js [`0f42d75`](https://github.com/Dragohm/travelling/commit/0f42d75b8c84ff53fd31785561ab23c726ec3f28)

#### [v2.18.0](https://github.com/Dragohm/travelling/compare/v2.17.3...v2.18.0)

> 22 August 2022

- Feature #60 edit user data with dot notation [`#22`](https://github.com/Dragohm/travelling/pull/22)
- Add user data sec tests [`6a1e809`](https://github.com/Dragohm/travelling/commit/6a1e8099f892d1d68fff5885fccafe3ea6b17d11)
- Add user data validation [`d3f6733`](https://github.com/Dragohm/travelling/commit/d3f67336ab65e4f5b8c5112a01fed8c738f9cd96)
- Other mr comment fix [`5a1a1ea`](https://github.com/Dragohm/travelling/commit/5a1a1ea33041aaa2bca068c0df269f02fe7e8267)
- MR comment update [`2046af5`](https://github.com/Dragohm/travelling/commit/2046af57e5c9ab87ce2c9408eb3d7c82e5aed5ac)
- fix stuff [`db48839`](https://github.com/Dragohm/travelling/commit/db488393a0359064d03c0736f818ceff75577ffc)
- Add endpoints [`cf27f51`](https://github.com/Dragohm/travelling/commit/cf27f511f0f41fe3611b35e88c0a39262d780984)
- Add tests [`3723929`](https://github.com/Dragohm/travelling/commit/3723929a408a3a425a152270b2c8d8919ff469f2)
- Add user prop object editing by value [`1b7c46d`](https://github.com/Dragohm/travelling/commit/1b7c46dab3e138a1dadc7a7e97e1196b431da5db)

#### [v2.17.3](https://github.com/Dragohm/travelling/compare/v2.17.2...v2.17.3)

> 26 July 2022

- Update user.js [`8cf0cca`](https://github.com/Dragohm/travelling/commit/8cf0ccac72ed4d1a7f3c3929f0964a4bc60a5a29)

#### [v2.17.2](https://github.com/Dragohm/travelling/compare/v2.17.1...v2.17.2)

> 11 July 2022

- fixes for node 16.15.1 and 18.4.0 [`#21`](https://github.com/Dragohm/travelling/pull/21)
- Update cookietoken.js [`8cd1323`](https://github.com/Dragohm/travelling/commit/8cd1323fa891a21ed7bb5facf62469409bbcb232)
- wrap in promise [`9139fed`](https://github.com/Dragohm/travelling/commit/9139fed26991b328e629d2dd20275e40fb6843cc)

#### [v2.17.1](https://github.com/Dragohm/travelling/compare/v2.17.0...v2.17.1)

> 11 July 2022

#### [v2.17.0](https://github.com/Dragohm/travelling/compare/v2.16.1...v2.17.0)

> 11 July 2022

- Fix #55 get users by group [`#20`](https://github.com/Dragohm/travelling/pull/20)
- Merge pull request #20 from Dragohm/fix-#55-get-users-by-group [`#55`](https://github.com/Dragohm/travelling/issues/55)
- group type comparison [`ddae09c`](https://github.com/Dragohm/travelling/commit/ddae09cb06f34640c72cb9dc7b358e1153a6a0bd)
- MR comment fixes [`8eb1e8c`](https://github.com/Dragohm/travelling/commit/8eb1e8c5d285cf32a23182a3c8270ef5ce17d2eb)
- Add tests. Update SDK [`d068a47`](https://github.com/Dragohm/travelling/commit/d068a47a3d99e72658618ed5e6fe5a50cde45666)

include/utils/config.js
const { resolve } = require('path');

require('dotenv').config({ path: resolve(process.cwd(), process.env.TRAVELLING_ENV || '.env') });

const misc = require('./misc');

const config = {
  port: misc.isSetDefault(Number(process.env.TRAVELLING_PORT), 443),
  ip: misc.isSetDefault(process.env.TRAVELLING_IP, '0.0.0.0'),
  key: misc.isSetDefault(process.env.TRAVELLING_KEY, __dirname + '/localhost.key'),
  cert: misc.isSetDefault(process.env.TRAVELLING_CERT, __dirname + '/localhost.csr'),
  serviceName: misc.isSetDefault(process.env.TRAVELLING_SERVICE_NAME, 'travelling'),
  cors: {
    enable: misc.isSetDefault(process.env.TRAVELLING_CORS_ENABLE, false),
    origin: misc.isSetDefault(process.env.TRAVELLING_CORS_HEADER_ORIGIN, null),
    methods: misc.isSetDefault(process.env.TRAVELLING_CORS_HEADER_METHODS, null),
    headers: misc.isSetDefault(process.env.TRAVELLING_CORS_HEADER_HEADERS, null),
    credentials: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_CORS_HEADER_CREDENTIALS), true),
    age: misc.isSetDefault(Number(process.env.TRAVELLING_CORS_HEADER_MAX_AGE), 3600)
  },
  https: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_HTTPS), true),
  misc: {
    bodyLimit: misc.isSetDefault(Number(process.env.TRAVELLING_MISC_BODY_LIMIT) * 1048576, 1048576 * 10), // 10MB default
    cloudflareIP: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_MISC_CLOUDFLAREIP), false),
    cloudflareDomainHeader: misc.isSetDefault(process.env.TRAVELLING_MISC_CLOUDFLARE_DOMAIN_HEADER, 'CF-Worker'),
    domainCustomHeader: misc.isSetDefault(process.env.TRAVELLING_MISC_DOMAIN_CUSTOM_HEADER, null),
    deniedRedirect: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_MISC_DENIED_REDIRECT), true),
    maxParamLength: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_MISC_MAX_PARAM_LENGTH), 100)
  },
  log: {
    enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_ENABLE), true),
    colors: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_COLORS), true),
    jsonoutput: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_JSONOUTPUT), false),
    appendFields: {
      version: {
        enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_APPEND_FIELDS_VERSION_ENABLE), true),
        label: misc.isSetDefault(process.env.TRAVELLING_LOG_APPEND_FIELDS_VERSION_LABEL, misc.getVersion())
      },
      host: {
        enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_APPEND_FIELDS_HOST_ENABLE), true),
        label: misc.isSetDefault(process.env.TRAVELLING_LOG_APPEND_FIELDS_HOST_LABEL, misc.getLocalIP())
      },
      app: {
        enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_APPEND_FIELDS_APP_ENABLE), true),
        label: misc.isSetDefault(
          process.env.TRAVELLING_LOG_APPEND_FIELDS_APP_LABEL,
          misc.isSetDefault(process.env.TRAVELLING_SERVICE_NAME, 'travelling')
        )
      },
      branch: {
        enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_APPEND_FIELDS_BRANCH_ENABLE), false),
        label: misc.isSetDefault(process.env.TRAVELLING_LOG_APPEND_FIELDS_BRANCH_LABEL, 'none')
      },
      environment: {
        enable: misc.isSetDefault(
          misc.stringToBool(process.env.TRAVELLING_LOG_APPEND_FIELDS_ENVIRONMENT_ENABLE),
          false
        ),
        label: misc.isSetDefault(process.env.TRAVELLING_LOG_APPEND_FIELDS_ENVIRONMENT_LABEL, 'production')
      }
    },
    level: misc.isSetDefault(process.env.TRAVELLING_LOG_LEVEL, 'info'),
    logger: misc.isSetDefault(process.env.TRAVELLING_LOG_LOGGER, 'wog'),
    requests: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_REQUESTS), true),
    unauthorizedAccess: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_UNAUTHORIZED_ACCESS), true),
    fastify: {
      logger: misc.isSetDefault(
        process.env.TRAVELLING_LOG_FASTIFY_LOGGER,
        misc.isSetDefault(process.env.TRAVELLING_LOG_LOGGER, 'wog')
      ),
      requestLogging: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_LOG_FASTIFY_LOGGER_REQUEST), true),
      requestIdHeader: misc.isSetDefault(
        misc.stringToBool(process.env.TRAVELLING_LOG_FASTIFY_LOGGER_REQ_ID_HEADER) ||
          process.env.TRAVELLING_LOG_FASTIFY_LOGGER_REQ_ID_HEADER,
        'travelling-req-id'
      ),
      requestIdLogLabel: misc.isSetDefault(
        misc.stringToBool(process.env.TRAVELLING_LOG_FASTIFY_LOGGER_REQ_ID_LOG_LABEL) ||
          process.env.TRAVELLING_LOG_FASTIFY_LOGGER_REQ_ID_LOG_LABEL,
        'travellingReqID'
      )
    }
  },
  audit: {
    create: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_AUDIT_CREATE_ENABLE), false)
    },
    edit: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_AUDIT_EDIT_ENABLE), false)
    },
    delete: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_AUDIT_DELETE_ENABLE), false)
    },
    view: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_AUDIT_VIEW_ENABLE), false)
    }
  },
  tracing: {
    enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_TRACING_ENABLE), false),
    instrumentations: misc.isSetDefault(
      process.env.TRAVELLING_TRACING_INSTRUMENTATIONS,
      resolve(__dirname, '../../include/server/tracing/instrumentations.js')
    ),
    processors: misc.isSetDefault(
      process.env.TRAVELLING_TRACING_PROCESSORS,
      resolve(__dirname, '../../include/server/tracing/processors.js')
    ),
    propagators: misc.isSetDefault(
      process.env.TRAVELLING_TRACING_PROPAGATORS,
      resolve(__dirname, '../../include/server/tracing/propagators.js')
    )
  },
  portal: {
    enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_PORTAL_ENABLE), true),
    path: misc.isSetDefault(
      process.env.TRAVELLING_PORTAL_PATH,
      `/${misc.isSetDefault(process.env.TRAVELLING_SERVICE_NAME, 'travelling')}/portal/`
    ),
    host: misc.isSetDefault(process.env.TRAVELLING_PORTAL_HOST, null),
    filePath: misc.isSetDefault(process.env.TRAVELLING_PORTAL_FILE_PATH, __dirname + '/../../client/dist'),
    logo: misc.isSetDefault(process.env.TRAVELLING_PORTAL_LOGO, __dirname + '/../../client/assets/logo.svg'),
    styles: misc.isSetDefault(process.env.TRAVELLING_PORTAL_STYLES, __dirname + '/../../client/assets/styles.css'),
    icon: misc.isSetDefault(process.env.TRAVELLING_PORTAL_ICON, __dirname + '/../../client/assets/favicon.ico'),
    webclient: {
      translation: {
        username: misc.isSetDefault(process.env.TRAVELLING_PORTAL_WEBCLIENT_TRANSLATION_USERNAME, 'Username'),
        password: misc.isSetDefault(process.env.TRAVELLING_PORTAL_WEBCLIENT_TRANSLATION_PASSWORD, 'Password'),
        email: misc.isSetDefault(process.env.TRAVELLING_PORTAL_WEBCLIENT_TRANSLATION_EMAIL, 'Email')
      }
    }
  },
  proxy: {
    timeout: misc.isSetDefault(Number(process.env.TRAVELLING_PROXY_TIMEOUT), 0),
    sendTravellingHeaders: misc.isSetDefault(
      misc.stringToBool(process.env.TRAVELLING_PROXY_SEND_TRAVELLING_HEADERS),
      false
    )
  },
  stats: {
    captureGroupRoutes: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_STATS_CAPTURE_GROUP_ROUTES), true)
  },
  redis: {
    enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_REDIS_ENABLE), false),
    url: misc.isSetDefault(process.env.TRAVELLING_REDIS_URL, 'redis://127.0.0.1:6379/'),
    events: {
      url: misc.isSetDefault(
        process.env.TRAVELLING_REDIS_EVENTS_URL,
        misc.isSetDefault(process.env.TRAVELLING_REDIS_URL, 'redis://127.0.0.1:6379/')
      )
    }
  },
  cookie: {
    session: {
      secret: misc.isSetDefault(process.env.TRAVELLING_COOKIE_SESSION_SECRET, null),
      expiration: misc.isSetDefault(Number(process.env.TRAVELLING_COOKIE_SESSION_EXPIRATION), 300) // seconds
    },
    token: {
      secret: misc.isSetDefault(process.env.TRAVELLING_COOKIE_TOKEN_SECRET, null),
      salt: misc.isSetDefault(process.env.TRAVELLING_COOKIE_TOKEN_SALT, null),
      expiration: misc.isSetDefault(Number(process.env.TRAVELLING_COOKIE_TOKEN_EXPIRATION), 30), // days
      checkable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_COOKIE_TOKEN_CHECKABLE), true) // Adds js accessible cookie (trav:ls) contianing expiration of trav:tok
    },
    domain: misc.isSetDefault(process.env.TRAVELLING_COOKIE_DOMAIN, null),
    security: {
      ipHijackProtection: misc.isSetDefault(
        misc.stringToBool(process.env.TRAVELLING_COOKIE_SECURITY_IP_HIJACK_PROTECTION),
        true
      )
    }
  },
  user: {
    isolateByDomain: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_USER_ISOLATE_BY_DOMAIN), false),
    username: {
      minchar: misc.isSetDefault(Number(process.env.TRAVELLING_USER_USERNAME_MINCHAR), 4),
      enabled: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_USER_USERNAME_ENABLED), true)
    },
    locked: {
      message: misc.isSetDefault(
        process.env.TRAVELLING_USER_LOCKED_MESSAGE,
        'Failed login attempts exceeded the limit. Contact your admin to get your account unlocked.'
      )
    }
  },
  password: {
    randomGenerationLength: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_RANDOM_GENERATION_LENGTH), 30),
    consecutive: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_PASSWORD_CONSECUTIVE), false),
    minchar: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_MINCHAR), 8),
    maxchar: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_MAXCHAR), ''),
    special: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_SPECIAL), 1),
    number: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_NUMBER), 1),
    lowercase: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_LOWERCASE), 1),
    uppercase: misc.isSetDefault(Number(process.env.TRAVELLING_PASSWORD_UPPERCASE), 1)
  },
  login: {
    maxLoginAttempts: misc.isSetDefault(Number(process.env.TRAVELLING_LOGIN_MAX_LOGIN_ATTEMPTS), 10)
  },
  token: {
    access: {
      expiration: misc.isSetDefault(Number(process.env.TRAVELLING_TOKEN_ACCESS_EXPIRATION), 1440) // minutes
    },
    code: {
      expiration: misc.isSetDefault(Number(process.env.TRAVELLING_TOKEN_CODE_EXPIRATION), 5), // minutes
      authorizeFlow: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_TOKEN_CODE_AUTHORIZE_FLOW), true)
    }
  },
  otp: {
    expiration: misc.isSetDefault(Number(process.env.TRAVELLING_OTP_EXPIRATION), 5 * 60) // seconds
  },
  pg: {
    url: misc.isSetDefault(process.env.TRAVELLING_DATABASE_URL, null),
    user: misc.isSetDefault(process.env.TRAVELLING_DATABASE_USER, null),
    password: misc.isSetDefault(process.env.TRAVELLING_DATABASE_PASSWORD, null),
    port: misc.isSetDefault(process.env.TRAVELLING_DATABASE_PORT, null),
    database: misc.isSetDefault(process.env.TRAVELLING_DATABASE_NAME, null),
    host: misc.isSetDefault(process.env.TRAVELLING_DATABASE_HOST, null),
    crypto: {
      implementation: misc.isSetDefault(
        process.env.TRAVELLING_PG_CRYPTO_IMPLEMENTATION,
        __dirname + '/cryptointerface.js'
      ),
      secret: misc.isSetDefault(process.env.TRAVELLING_PG_CRYPTO_IMPLEMENTATION_SECRET, null),
      salt: misc.isSetDefault(process.env.TRAVELLING_PG_CRYPTO_IMPLEMENTATION_SALT, null),
      encryptUserData: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_PG_CRYPTO_ENCRYPT_USER_DATA), false)
    }
  },
  email: {
    validation: {
      external: {
        enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_VALIDATION_EXTERNAL_ENABLE), false),
        endpoint: misc.isSetDefault(process.env.TRAVELLING_EMAIL_VALIDATION_EXTERNAL_ENDPOINT, null),
        emailInEndpoint: misc.isSetDefault(
          misc.stringToBool(process.env.TRAVELLING_EMAIL_VALIDATION_EXTERNAL_EMAIL_IN_ENDPOINT),
          true
        ),
        emailInBody: misc.isSetDefault(
          misc.stringToBool(process.env.TRAVELLING_EMAIL_VALIDATION_EXTERNAL_EMAIL_IN_BODY),
          false
        ),
        method: misc.isSetDefault(process.env.TRAVELLING_EMAIL_VALIDATION_EXTERNAL_METHOD, 'GET')
      }
    },
    from: misc.isSetDefault(process.env.TRAVELLING_EMAIL_FROM, null),
    recovery: {
      expiration: misc.isSetDefault(Number(process.env.TRAVELLING_EMAIL_RECOVERY_EXPIRATION), 900) // seconds
    },
    activation: {
      expiration: misc.isSetDefault(Number(process.env.TRAVELLING_EMAIL_ACTIVATION_EXPIRATION), 86400) // seconds
    },
    test: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_TEST_ENABLE), false)
    },
    smtp: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_SMTP_ENABLE), false),
      host: misc.isSetDefault(process.env.TRAVELLING_EMAIL_SMTP_HOST, '127.0.0.1'),
      port: misc.isSetDefault(process.env.TRAVELLING_EMAIL_SMTP_PORT, 465),
      secure: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_SMTP_SECURE), true), // use TLS
      auth: {
        user: misc.isSetDefault(process.env.TRAVELLING_EMAIL_SMTP_AUTH_USER, null),
        pass: misc.isSetDefault(process.env.TRAVELLING_EMAIL_SMTP_AUTH_PASSWORD, null)
      },
      tls: {
        // do not fail on invalid certs
        rejectUnauthorized: misc.isSetDefault(
          misc.stringToBool(process.env.TRAVELLING_EMAIL_SMTP_TLS_REJECT_UNAUTHORIZED),
          true
        )
      }
    },
    aws: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_AWS_ENABLE), false),
      config: misc.isSetDefault(process.env.TRAVELLING_EMAIL_AWS_CONFIG, null)
    },
    rest: {
      enable: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_REST_ENABLE), false),
      passwordRecoveryEndpoint: misc.isSetDefault(process.env.TRAVELLING_EMAIL_REST_PASSWORD_RECOVERY_ENDPOINT, null),
      activationEndpoint: misc.isSetDefault(process.env.TRAVELLING_EMAIL_REST_ACTIVATION_ENDPOINT, null),
      lockedEndpoint: misc.isSetDefault(process.env.TRAVELLING_EMAIL_REST_LOCKED_ENDPOINT, null),
      welcomeEndpoint: misc.isSetDefault(process.env.TRAVELLING_EMAIL_REST_WELCOME_ENDPOINT, null)
    },
    send: {
      onLocked: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_SEND_ON_LOCKED), true),
      onNewUser: misc.isSetDefault(misc.stringToBool(process.env.TRAVELLING_EMAIL_SEND_ON_NEW_USER), false)
    },
    template: {
      passwordResetBody: misc.isSetDefault(
        process.env.TRAVELLING_EMAIL_RESET_PASSWORD_TEMPLATE_BODY,
        __dirname + '/../../templates/email-reset-password-body.html'
      ),
      passwordResetSubject: misc.isSetDefault(
        process.env.TRAVELLING_EMAIL_RESET_PASSWORD_TEMPLATE_SUBJECT,
        __dirname + '/../../templates/email-reset-password-subject.html'
      ),
      activationBody: misc.isSetDefault(
        process.env.TRAVELLING_EMAIL_ACTIVATION_TEMPLATE_BODY,
        __dirname + '/../../templates/email-activation-body.html'
      ),
      activationSubject: misc.isSetDefault(
        process.env.TRAVELLING_EMAIL_ACTIVATION_TEMPLATE_SUBJECT,
        __dirname + '/../../templates/email-activation-subject.html'
      ),
      welcomeBody: misc.isSetDefault(
        process.env.TRAVELLING_EMAIL_WELCOME_TEMPLATE_BODY,
        __dirname + '/../../templates/email-welcome-body.html'
      ),
      welcomeSubject: misc.isSetDefault(
        process.env.TRAVELLING_EMAIL_WELCOME_TEMPLATE_SUBJECT,
        __dirname + '/../../templates/email-welcome-subject.html'
      )
    }
  },
  registration: {
    requireEmailActivation: misc.isSetDefault(
      misc.stringToBool(process.env.TRAVELLING_REGISTRATION_REQUIRE_EMAIL_ACTIVATION),
      false
    ),
    requireManualActivation: misc.isSetDefault(
      misc.stringToBool(process.env.TRAVELLING_REGISTRATION_REQUIRE_MANUAL_ACTIVATION),
      false
    )
  }
};

module.exports = config;
